ocaml/xapi/livepatch.ml

```diff
@@ -0,0 +1,343 @@
+    Option.(value (map (fun x -> `String x) o) ~default:`Null)
+  in
+  `Assoc
+    [
+      ("component", `String (string_of_component lp.component))
+    ; ("base_build_id", `String lp.base_build_id)
+    ; ("base_version", `String lp.base_version)
+    ; ("base_release", `String lp.base_release)
+    ; ("to_version", option_to_json lp.to_version)
+    ; ("to_release", option_to_json lp.to_release)
+    ]
+
+let of_json js =
+  try
+    let open Yojson.Basic.Util in
+    {
+      component= member "component" js |> to_string |> component_of_string
+    ; base_build_id= member "base_build_id" js |> to_string
+    ; base_version= member "base_version" js |> to_string
+    ; base_release= member "base_release" js |> to_string
+    ; to_version= member "to_version" js |> to_string_option
+    ; to_release= member "to_release" js |> to_string_option
+    }
+  with e ->
+    let msg = "Can't construct a livepatch from json" in
+    error "%s: %s" msg (ExnHelper.string_of_exn e) ;
+    raise Api_errors.(Server_error (internal_error, [msg]))
+
+let get_latest_livepatch lps =
+  List.map (fun (_, _, t_v, t_r) -> (t_v, t_r)) lps
+  |> Rpm.get_latest_version_release
+  |> function
+  | Some (v, r) ->
+      List.find_opt (fun (_, _, t_v, t_r) -> t_v = v && t_r = r) lps
+  | None ->
+      None
+
+module KernelLivePatch = struct
+  let get_running_livepatch () =
+    let r =
+      (* This expects the kernel livepatch module name should be built as:
+       * livepatch_<base_version>__<base_release>__<to_version>__<to_releaes>
+       * The "." in <base_version>, <base_release>, <to_version>, and <to_releaes>
+       * should be replaced with "_".
+       * For example:
+       * livepatch_4_19_19__8_0_20__4_19_19__8_0_22
+       *)
+      Re.Posix.compile_pat
+        {|^[ ]*livepatch_([^ \[]+)__([^ \[]+)__([^ \[]+)__([^ \[]+).*$|}
+    in
+    Helpers.call_script !Xapi_globs.kpatch_cmd ["list"]
+    |> Astring.String.cuts ~sep:"\n"
+    |> List.fold_left
+         (fun (acc, acc_flag) line ->
+           match acc_flag with
+           | true ->
+               (acc, true)
+           | false ->
+               if
+                 Astring.String.is_prefix ~affix:"Installed patch modules:" line
+               then
+                 (acc, true)
+               else
+                 (line :: acc, false)
+         )
+         ([], false)
+    |> (fun (x, _) -> x)
+    |> List.filter_map (fun line ->
+           let replace s =
+             Astring.String.cuts ~sep:"-" s |> Astring.String.concat ~sep:"."
+           in
+           match Re.exec_opt r line with
+           | Some groups ->
+               Some
+                 ( Re.Group.get groups 1 |> replace
+                 , Re.Group.get groups 2 |> replace
+                 , Re.Group.get groups 3 |> replace
+                 , Re.Group.get groups 4 |> replace
+                 )
+           | None ->
+               None
+       )
+    |> get_latest_livepatch
+
+  let get_base_build_id () =
```

Comment:
What is this code doing? It seems quite repetitive and using low-level operations. Parsing binary data could be easier using `Angstrom` which has operations to read bytes and Big Endian, Small Endian numbers.

Reply:
It is going to read ELF notes from file `/sys/kernel/notes` to find out the running kernel's build id.
The file content is organized as some consecutive sections. For each section, at beginning of it, it's a header structure described by `headers_struct`. And the length of a section is determined by its headers. This function needs to find a specific section and read build id in it.

If `Angstrom` works for this in better way, I will look into it. Thanks :)

Reply:
* https://github.com/lindig/fit/blob/main/lib/fit.ml#L213-L222

For example, this is code I wrote to read the (binary) header of a FIT file. There is more complex code in that file but it gives you a flavor how to read bytes, multi-byte numbers, and strings.

Reply:
There's another example of using angstrom for decoding the output of a command, it might prove useful as well: https://github.com/xapi-project/xen-api/blob/47fae74032aa6ade0fc12e867c530eaf2a96bf75/ocaml/xapi/bios_strings.ml

Looking at the documentation of angstrom was useful when I developed the module: https://docs.ocaml.pro/docs/LIBRARY.angstrom@angstrom.0.15.0/Angstrom/index.html

Reply:
Hi @lindig @psafont 
Thanks for your pointers. I've reworked this with `Angstrom` which looks very great.